Add deprecated removed_in_version and deprecated_aliases version tests, allow to enable such tests via flag

felixfontein opened the pull request:
##### SUMMARY
Fixes #66919.

Makes sure that `removed_in_version` and `deprecated_aliases` are compared to the current Ansible version.

This is a WIP since there is currently no good mechanism to only run this when a new Ansible version is released. Idea to solve this: add `ignore.txt` entries for current values and create issues; `ignore.txt` has to be updated when Ansible's version number is bumped (i.e. after a release). The same should probably be done for the pylint deprecation test (which is currently controlled with a pylint config file).

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
ansible-test

felixfontein committed:
Add validate-module test for deprecated removed_in_version.

felixfontein commented:
The test failure is intentional; https://app.shippable.com/github/ansible/ansible/runs/157507/5/tests lists all affected options. There are 737 options which should have been removed for 2.9 or 2.10. I think quite some of them might come from doc fragments though, so the situation is not *that* bad :)

felixfontein committed:
Also cover deprecated_aliases.

felixfontein commented:
I've added a check for version numbers in `deprecated_aliases`.

felixfontein commented:
The new check didn't find any error with `deprecated_aliases`, the list in https://app.shippable.com/github/ansible/ansible/runs/157736/5/tests is still the same.

felixfontein commented:
An idea for deprecation handling which also works well for collections: how about making the `ansible-deprecated-version` error include the Ansible version? I.e. change the error to `ansible-deprecated-version-2.10` (for now).

That way, one can put the errors into `ignore.txt` after a Ansible version bump until they are fixed, and if they aren't fixed by the next version bump, you'll get a nice reminder by the fact that `ansible-deprecated-version-2.old` is no longer needed, but `ansible-deprecated-version-2.new` is.

That way, the current method (disable the current existing `ansible-deprecated-version` check which runs via pylint) can be stopped, i.e. the test can be reenabled, and current problems be put into ignore.txt, without them (or similar ones for newer versions) getting forgotten forever.

The version number included in `ansible-deprecated-version-2.x` could also be the one used in the `deprecate` call / `removed_in_version` field / `deprecated_aliases` field. That way, there's no need to update existing `ignore.txt` entries when a new Ansible version is released.

felixfontein commented:
Plan: only run these checks outside collections (`if not self.collection:` should do that), and enable pylint `module.deprecate()` sanity check also for non-collections.

felixfontein committed:
Run all ansible-deprecated-version tests, but not in collections.

felixfontein commented:
@jctanner does this now look better to you?

felixfontein commented:
TODO next: add a switch which enables this test (and also check for removal date once #68177 is there)

felixfontein committed:
Allow to enable deprecation version validation with ansible-test sanity CLI flag '--validate-deprecation'.

felixfontein commented:
There is a now a CLI flag for `ansible sanity` called `--validate-deprecation` which enables the deprecation version validation checks - both the existing one (pylint for `module.deprecate()` calls) and the new ones (for `removed_in_version` and `deprecated_aliases` in argspec).

Sample error output of `ansible-test sanity --test validate-modules --validate-deprecation lib/ansible/modules/network/dellos10/dellos10_command.py`:
```
ERROR: Found 8 validate-modules issue(s) which need to be resolved:
ERROR: lib/ansible/modules/network/dellos10/dellos10_command.py:0:0: ansible-deprecated-version: Argument 'auth_pass' in argument_spec has a deprecated removed_in_version '2.9', i.e. the version is less than or equal to the current version of Ansible (2.10.0.dev0)
ERROR: lib/ansible/modules/network/dellos10/dellos10_command.py:0:0: ansible-deprecated-version: Argument 'authorize' in argument_spec has a deprecated removed_in_version '2.9', i.e. the version is less than or equal to the current version of Ansible (2.10.0.dev0)
ERROR: lib/ansible/modules/network/dellos10/dellos10_command.py:0:0: ansible-deprecated-version: Argument 'host' in argument_spec has a deprecated removed_in_version '2.9', i.e. the version is less than or equal to the current version of Ansible (2.10.0.dev0)
ERROR: lib/ansible/modules/network/dellos10/dellos10_command.py:0:0: ansible-deprecated-version: Argument 'password' in argument_spec has a deprecated removed_in_version '2.9', i.e. the version is less than or equal to the current version of Ansible (2.10.0.dev0)
ERROR: lib/ansible/modules/network/dellos10/dellos10_command.py:0:0: ansible-deprecated-version: Argument 'port' in argument_spec has a deprecated removed_in_version '2.9', i.e. the version is less than or equal to the current version of Ansible (2.10.0.dev0)
ERROR: lib/ansible/modules/network/dellos10/dellos10_command.py:0:0: ansible-deprecated-version: Argument 'ssh_keyfile' in argument_spec has a deprecated removed_in_version '2.9', i.e. the version is less than or equal to the current version of Ansible (2.10.0.dev0)
ERROR: lib/ansible/modules/network/dellos10/dellos10_command.py:0:0: ansible-deprecated-version: Argument 'timeout' in argument_spec has a deprecated removed_in_version '2.9', i.e. the version is less than or equal to the current version of Ansible (2.10.0.dev0)
ERROR: lib/ansible/modules/network/dellos10/dellos10_command.py:0:0: ansible-deprecated-version: Argument 'username' in argument_spec has a deprecated removed_in_version '2.9', i.e. the version is less than or equal to the current version of Ansible (2.10.0.dev0)
```
When called without `--validate-deprecation`, validation passes.

felixfontein committed:
Allow to validate collection deprecations (when collection version is known)

felixfontein committed:
Parse galaxy.yml for collections and provide collection version.

felixfontein committed:
Only provide options for deprecated pylint plugin when plugin is loaded.

felixfontein committed:
Remove debug code.

felixfontein committed:
Linting.

felixfontein committed:
More linting.

felixfontein committed:
Make sure constructor args are passed through.

mattclay commented on test/lib/ansible_test/_internal/cli.py:
```diff
@@ -527,6 +527,11 @@ def key_value_type(value):  # type: (str) -> t.Tuple[str, str]
+                        action='store_true',
+                        help='Compare deprecation versions to current Ansible '
+                             'version resp. collection version')
+
```
Validation should always be enabled. Otherwise the validity of sanity ignores will be dependent on CLI options, which is something we work very hard to avoid.

felixfontein replied:
In the meeting, it was explicitly wished that this is not enabled by default.

https://meetbot.fedoraproject.org/ansible-meeting/2020-03-17/ansible_core_public_irc_meeting_https:github.comansiblecommunityissues528.2020-03-17-19.01.log.html
See timestamp 19:34:00 and later.

I personally would like it to be enabled by default as well, but especially when doing deprecations by date (#68177), this will result in situations where tests suddenly fail from one day to the next. I think that's the main motivation to not enable it by default.

mattclay replied:
Deprecation by date should be opt-in, but the existing deprecation by version can remain always-on.

However, I'd like to change how the opt-in behavior is implemented. In ansible-test we should declare the error code(s) that are opt-in for a given test. By default ansible-test would then filter out these codes from test results, as well as prevent them from being ignored to maintain the integrity of test results. When running tests with optional codes, those codes would always be tested, which would help ensure they function correctly even when not shown.

This of course will require use of separate error codes for deprecation by date and deprecation by version.

We can add an `--enable-optional` argument to ansible-test that enables reporting of the optional codes.

On a related note, the pylint configuration needs to be updated to stop ignoring the deprecation errors.

This opt-in test behavior will also be useful for https://github.com/ansible/ansible/pull/64623

felixfontein replied:
@mattclay there's currently no always-on deprecation by version check. It is always-off, and the only way to turn it on is to edit ansible-test internals. Or do you mean the current state of this PR?

mattclay replied:
When I looked at the pylint config for ansible (not collections), the `git blame` showed I added the entry to disable `ansible-deprecated-version` and I was thinking I'd added it by mistake when splitting the config between ansible and collections. I see now that wasn't the case.

That being said, the deprecation by version could be always-on since we have ignores. I think it would be more useful as such. For the reasons already stated, the deprecation by date cannot be.

mattclay commented on test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py:
```diff
@@ -19,7 +19,8 @@
               "Used when a call to Display.deprecated specifies a version "
-              "less than or equal to the current version of Ansible",
+              "less than or equal to the current version of Ansible resp. "
+              "this collection",
```
Instead of making this message apply to both Ansible and collections, why not have two error codes/messages, one for each scenario?

felixfontein replied:
Implemented in e86cae662ded639828c8828aae177b47a1244a74.

mattclay commented on test/lib/ansible_test/_internal/provider/layout/__init__.py:
```diff
@@ -169,6 +170,19 @@ def __init__(self,
+        except IOError as exc:
+            raise ApplicationError('Error while reading galaxy.yml: {0}'.format(exc))
+
+        self.version = galaxy.get('version')
```
This should work on both source and published collections, which means it will need to support reading from both `galaxy.yml` and `MANIFEST.json`. If the version lookup fails for any reason, it should continue with a warning. When the pylint plugin or validate-modules attempt to check a deprecated version, at that point they should generate an error if the version was not provided.

This will ensure that `ansible-test` can function on a local collection with no version, or corrupted metadata. The only tests which will report errors are the ones trying to use that version information at the time of use. If no version checking needs to be done, lack of a version to check against shouldn't be an error.

Also, the version lookup in `ansible-test` itself should be lazy-loaded, and deferred to a subprocess.

For an example, see:

https://github.com/ansible/ansible/blob/0651e1c25fab26c33130128774772444957fc524/test/lib/ansible_test/_internal/ansible_util.py#L190

That way `ansible-test` can be run from source, such as from GitHub, to test a collection using delegation (`--docker` for example) and not require `yaml` to be installed to avoid a spurious warning. It also makes sure that the correct python is used to import `yaml` when using the `--python` option, as that interpreter may be different than the one used to run `ansible-test`.

felixfontein replied:
I've added support for MANIFEST.json in 3e51422b049453b23b9c2fb668c912f006514d18, and then moved metadata collection to a subprocess in eea0120ecfafa9ae563fc99688b6c590889c8074 as you suggested. I put the resulting function also in ansible_util.py. Do you prefer another place?

BTW, I noticed that `ansible-test sanity` doesn't really works in galaxy-installed collections. I always get `WARNING: All targets skipped.` and no test is run. Happens both with this branch and with `devel`.

mattclay replied:
@felixfontein Can you open an issue for that?

felixfontein replied:
Done (#68499)

felixfontein committed:
Add new error code and message for collections.

felixfontein committed:
Read MANIFEST.json to obtain collection version. Make non-existance of galaxy.yml or MANIFEST.json not an error.

felixfontein committed:
Move collection metadata collection to subprocess, and only run when needed.

felixfontein committed:
Forgot to remove debug code.

felixfontein committed:
Linting. Adding warning if neither galaxy.yml nor MAINFEST.json is found.

felixfontein committed:
Remove no longer necessary import.

felixfontein committed:
Always enable deprecation version checking.

felixfontein committed:
Mark deprecation version error codes as optional error codes, and allow to enable them with --enable-optional-errors.

felixfontein commented:
/rebuild

felixfontein commented:
/rebuild_failed

felixfontein commented:
(rebased to resolved conflicts)

felixfontein committed:
First source of truth is MANIFEST.json, only then comes galaxy.yml.

felixfontein commented:
@s-hertel [suggested](https://github.com/ansible/ansible/pull/69291#discussion_r421672339) to prefer `MANIFEST.json` over `galaxy.yml`. The latest commit 71c4471b44c4e0bdf65ea2bdf1f5bf6e2329979c addresses that.

felixfontein committed:
Forgot to adjust second ModuleValidator instantiation.

mattclay committed:
Review changes.

mattclay commented on test/lib/ansible_test/_data/collection_detail.py:
```diff
@@ -0,0 +1,69 @@
+        collection_info = manifest.get('collection_info') or dict()
+
+        result = dict(
+            version=collection_info.get('version'),
```
I just realized that the version lookup here (and below) should be passing the version through `LooseVersion` to make sure it's valid. Otherwise we'll get a traceback later when passing a bogus version to `pylint` or `validate-modules`.

felixfontein replied:
That reminds me that for collection versions, we should rather use `semantic_version.Version`  from https://pypi.org/project/semantic-version/ instead of `LooseVersion`.

felixfontein committed:
Return empty dictionary in case no metadata was found.

felixfontein committed:
Validating parsed versions.

Using regular expressions instead of semantic_version.Version() since semantic_version
is not necessarily installed.

felixfontein committed:
Check for invalid versions.

felixfontein committed:
Use semantic_version.Version instead of LooseVersion to compare collection versions.

felixfontein commented:
The latest commit (71277d9c44a5f3624e6524c80a72806428c6e63a) uses `semantic_version.Version` instead of `LooseVersion` to compare collection versions.

This has the side-effect that probably most existing collections will have a lot new errors, since every two-component version (like `2.14`) cannot be parsed (since they don't satisfy the semantic versioning spec), thus causing new sanity errors.

On the other hand, this is good IMO, since these values need to be updated anyway as they should reflect to collection version numbers, and not to ansible-base version numbers.

felixfontein commented:
@gundalow any opinions on this?

felixfontein committed:
Linting.

mattclay commented on test/lib/ansible_test/_data/collection_detail.py:
```diff
@@ -57,6 +57,8 @@ def main():
     try:
         result = read_manifest_json(collection_path) or read_galaxy_yml(collection_path)
+        if result is None:
+            result = dict()
```
A less verbose way to write this:

```suggestion
        result = read_manifest_json(collection_path) or read_galaxy_yml(collection_path) or {}
```

felixfontein replied:
Thanks, fixed it.

mattclay commented on test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py:
```diff
@@ -6,6 +6,12 @@
+    HAS_SEMANTIC_VERSION = True
+except ImportError as dummy:
+    HAS_SEMANTIC_VERSION = False
+
```
Since `semantic_version` is listed in the requirements I don't think we should be checking for it here. Just let it fail as it would for any other missing requirement.

felixfontein committed:
Less verbose.

mattclay commented on test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:
```diff
@@ -32,10 +32,17 @@
+    import semantic_version
+    HAS_SEMANTIC_VERSION = True
+except Exception as e:
+    HAS_SEMANTIC_VERSION = False
```
Since `semantic_version` is listed in the requirements I don't think we should be checking for it here. Just let it fail as it would for any other missing requirement.

Not special casing this will also prevent inconsistency in the reported errors. The errors reported should be consistent between environments, which means tests either need to fail (traceback) due to missing dependencies, or be skipped entirely.

mattclay commented on test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:
```diff
@@ -1450,6 +1468,72 @@ def _validate_argument_spec(self, docs, spec, kwargs, context=None, last_context
+                                code='ansible-deprecated-version',
+                                msg=msg,
+                            )
+                    except ValueError as dummy:
```
Since `dummy` isn't used don't include it:

```suggestion
                    except ValueError:
```

mattclay commented on test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:
```diff
@@ -1450,6 +1468,72 @@ def _validate_argument_spec(self, docs, spec, kwargs, context=None, last_context
+                                    code='ansible-deprecated-version',
+                                    msg=msg,
+                                )
+                        except ValueError as dummy:
```
Since `dummy` isn't used don't include it:

```suggestion
                    except ValueError:
```

felixfontein replied:
Fixed both.

felixfontein committed:
Assume semantic_version is installed.

felixfontein committed:
Remove unneeded variable.

mattclay commented on test/lib/ansible_test/_data/requirements/sanity.txt:
```diff
@@ -4,6 +4,7 @@ pycodestyle ; python_version >= '3.5' # only used on python 3.5+
 pylint ; python_version >= '3.5' and python_version < '3.9' # only used on python 3.5+ and does not yet support python 3.9
 pyyaml
 rstcheck ; python_version >= '3.5' # only used on python 3.5+
+semantic_version
```
This is only used for tests that require Python 3.5+

```suggestion
semantic_version ; python_version >= '3.5' # only used on python 3.5+
```

felixfontein committed:
Update test/lib/ansible_test/_data/requirements/sanity.txt

Co-authored-by: Matt Clay <matt@mystile.com>

mattclay commented on test/lib/ansible_test/_data/requirements/sanity.txt:
```diff
@@ -4,6 +4,7 @@ pycodestyle ; python_version >= '3.5' # only used on python 3.5+
 pylint ; python_version >= '3.5' and python_version < '3.9' # only used on python 3.5+ and does not yet support python 3.9
 pyyaml
 rstcheck ; python_version >= '3.5' # only used on python 3.5+
+semantic_version ; python_version >= '3.5' # only used on python 3.5+
```
I just realized we should be using the same `SemanticVersion` code that `ansible-galaxy` is using:

https://github.com/ansible/ansible/blob/98fbd1b72f501fbdee8388e95b02c2582717971a/lib/ansible/utils/version.py#L128

felixfontein replied:
Interesting, I didn't knew Ansible had it's own semantic version code. Anyway, while looking over it I found a bug I think. I'll adjust this PR to use the class and then I'll create a PR to fix the bug.

felixfontein replied:
Bugfix in #69395.

felixfontein committed:
Use Ansible's own SemanticVersion instead of semantic_version.Version.

mattclay commented:
This PR also needs a changelog fragment.

felixfontein committed:
Add changelog.

felixfontein commented:
/rebuild_failed

felixfontein committed:
Separate errors for collection versions than for Ansible versions.

felixfontein committed:
Add documentation.

felixfontein commented:
I updated the PR to use different error codes for collection version numbers than for Ansible version numbers, so that it is more obvious to module authors what is wrong. I've also updated the list of validate-modules error codes.

Finally, I've talked to @gundalow on IRC. It is OK that existing `removed_in_version` will be flagged in collections, since these are currently wrong anway (refering to Ansible versions instead of collection versions). Once this is merged, I'll write an announcement for this so that collection authors will know.

felixfontein commented:
/rebuild_failed

felixfontein committed:
Update ignore.txt.

felixfontein commented:
@mattclay thanks a lot for all the reviewing, feedback, and merging!
